StellarDotnetSdk.Tests/WebAuthenticationTest.cs

```diff
@@ -1691,60 +1719,193 @@ public void TestReadChallengeTransactionExpiredTimeBounds()
+        var now = DateTimeOffset.Now;
+        // Time bounds start 20 seconds before 'now'
+        var validFrom = now.AddSeconds(-20);
+        var validFor = TimeSpan.FromSeconds(10);
+
+        var transaction = _BuildAndSignChallengeTransaction(
+            validFrom,
+            validFor);
+
+        // Grace time bounds
+        var graceValidFrom = validFrom.AddSeconds(-WebAuthentication.GracePeriod);
+
+        // Act & Assert
+        Assert.ThrowsException<InvalidWebAuthenticationException>(() =>
+        {
+            _ReadChallengeTransaction(
+                transaction,
+                // Start 1 second before the min time bounds - grace period
+                graceValidFrom.AddSeconds(-1));
+        }, "Challenge transaction expired");
+    }
+
+    [TestMethod]
+    public void TestReadChallengeTransactionWithOutOfLowerBoundButWithinGracePeriod()
+    {
+        // Arrange
+        var now = DateTimeOffset.Now;
+        // Time bounds start 20 seconds before 'now'
+        var validFrom = now.AddSeconds(-20);
+        var validFor = TimeSpan.FromSeconds(10);
+
+        var transaction = _BuildAndSignChallengeTransaction(
+            validFrom,
+            validFor);
+
+        // Grace time bounds
+        var graceValidFrom = validFrom.AddSeconds(-WebAuthentication.GracePeriod);
+
+        // Act & Assert
+        var result = _ReadChallengeTransaction(
+            transaction,
+            // Start 1 second after the min time bounds - grace period
+            graceValidFrom.AddSeconds(1));
+        Assert.AreEqual(_clientKeypair.AccountId, result);
+    }
+
+    [TestMethod]
+    public void TestReadChallengeTransactionWithOutOfUpperBound()
+    {
+        // Arrange
+        var now = DateTimeOffset.Now;
+        // Time bounds start 20 seconds before 'now'
+        var validFrom = now.AddSeconds(-20);
+        var validFor = TimeSpan.FromSeconds(10);
+
+        var transaction = _BuildAndSignChallengeTransaction(
+            validFrom,
+            validFor);
+
+        // Grace time bounds
+        var graceValidFor = validFor + TimeSpan.FromSeconds(WebAuthentication.GracePeriod);
+        var graceValidTill = validFrom.Add(graceValidFor);
+
+        // Act & Assert
+        Assert.ThrowsException<InvalidWebAuthenticationException>(() =>
+        {
+            _ReadChallengeTransaction(
+                transaction,
+                // Start 1 second after the max time bounds + grace period
+                graceValidTill.AddSeconds(1));
+        }, "Challenge transaction expired");
+    }
+
+    [TestMethod]
+    public void TestReadChallengeTransactionWithOutOfUpperBoundButWithinGracePeriod()
+    {
+        // Arrange
+        var now = DateTimeOffset.Now;
+        // Time bounds start 20 seconds before 'now'
+        var validFrom = now.AddSeconds(-20);
+        var validFor = TimeSpan.FromSeconds(10);
+
+        var transaction = _BuildAndSignChallengeTransaction(
+            validFrom,
```

Comment:
Not important, but for function calls that are short enough to occupy 1 line, I'd prefer to keep them in 1 line.

Reply:
Alright, let's talk about this another time a bit more.